File: hdf-converters/shared/go/hdfversion/hdf_version_test.go
package hdfversion

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/xeipuuv/gojsonschema"
)

func convertersDir() string {
	_, thisFile, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(thisFile), "..", "..", "..", "converters")
}

func legacyhdfFixture(t *testing.T, name string) []byte {
	t.Helper()
	path := filepath.Join(convertersDir(), "legacyhdf-to-hdf", "fixtures", "input", name)
	data, err := os.ReadFile(path)
	require.NoError(t, err)
	return data
}

// Version identifiers: v2 = legacy Heimdall schema (profiles/platform),
// v3 = modern hdf-libs schema (baselines/components). There is no HDF v1
// transform (v1 = raw InSpec; see NormalizeVersion).

func TestTransformHDF_LegacyToModern(t *testing.T) {
	legacyInput := legacyhdfFixture(t, "minimal.json")

	output, _, err := TransformHDF(legacyInput, LegacyVersion, ModernVersion)
	require.NoError(t, err)
	require.NotEmpty(t, output)

	// Output should be valid JSON with modern (v3) structure.
	var modern map[string]any
	require.NoError(t, json.Unmarshal(output, &modern))
	assert.Contains(t, modern, "baselines", "modern output should have baselines")
	assert.Contains(t, modern, "components", "modern output should have components")
	// Should NOT have legacy fields.
	assert.NotContains(t, modern, "profiles", "modern output should not have profiles")
	assert.NotContains(t, modern, "platform", "modern output should not have platform")
}

func TestTransformHDF_ModernToLegacy(t *testing.T) {
	// First get a modern document by upgrading a legacy one.
	legacyInput := legacyhdfFixture(t, "minimal.json")
	modernOutput, _, err := TransformHDF(legacyInput, LegacyVersion, ModernVersion)
	require.NoError(t, err)

	// Now downgrade back to the legacy shape.
	legacyOutput, _, err := TransformHDF(modernOutput, ModernVersion, LegacyVersion)
	require.NoError(t, err)
	require.NotEmpty(t, legacyOutput)

	// Output should have legacy (v2) structure.
	var legacy map[string]any
	require.NoError(t, json.Unmarshal(legacyOutput, &legacy))
	assert.Contains(t, legacy, "profiles", "legacy output should have profiles")
	assert.Contains(t, legacy, "platform", "legacy output should have platform")
	assert.Contains(t, legacy, "statistics", "legacy output should have statistics")
	// Should NOT have modern fields.
	assert.NotContains(t, legacy, "baselines", "legacy output should not have baselines")
	assert.NotContains(t, legacy, "components", "legacy output should not have components")
}

func TestTransformHDF_SameVersion(t *testing.T) {
	legacyInput := legacyhdfFixture(t, "minimal.json")

	// Same version should return input unchanged.
	output, _, err := TransformHDF(legacyInput, LegacyVersion, LegacyVersion)
	require.NoError(t, err)
	assert.JSONEq(t, string(legacyInput), string(output))
}

func TestTransformHDF_UnknownTransform(t *testing.T) {
	// "1" is not a transform key — raw InSpec is not a distinct schema version.
	_, _, err := TransformHDF([]byte(`{}`), "1", ModernVersion)
	require.Error(t, err)
	assert.Contains(t, err.Error(), "no HDF transform")
}

func TestTransformHDF_InvalidJSON(t *testing.T) {
	_, _, err := TransformHDF([]byte(`not json`), LegacyVersion, ModernVersion)
	require.Error(t, err)
}

func TestTransformHDF_RoundTrip(t *testing.T) {
	legacyInput := legacyhdfFixture(t, "minimal.json")

	// legacy → modern → legacy should preserve core fields.
	modern, _, err := TransformHDF(legacyInput, LegacyVersion, ModernVersion)
	require.NoError(t, err)

	legacyAgain, _, err := TransformHDF(modern, ModernVersion, LegacyVersion)
	require.NoError(t, err)

	var original map[string]any
	var roundTripped map[string]any
	require.NoError(t, json.Unmarshal(legacyInput, &original))
	require.NoError(t, json.Unmarshal(legacyAgain, &roundTripped))

	origProfiles, _ := original["profiles"].([]any)
	rtProfiles, _ := roundTripped["profiles"].([]any)
	assert.Equal(t, len(origProfiles), len(rtProfiles), "profile count should survive round-trip")
}

func TestDowngradeV3ToV2_FlattensAmendments(t *testing.T) {
	_, thisFile, _, _ := runtime.Caller(0)
	input, err := os.ReadFile(filepath.Join(filepath.Dir(thisFile), "testdata", "modern_with_amendments.json"))
	require.NoError(t, err)

	out, warnings, err := TransformHDF(input, ModernVersion, LegacyVersion)
	require.NoError(t, err)

	var legacy map[string]any
	require.NoError(t, json.Unmarshal(out, &legacy))

	// Incidental gap C: top-level version reconstructed from the source tool.
	assert.Equal(t, "5.22.65", legacy["version"])

	// The downgraded document must satisfy the InSpec exec-json schema Heimdall loads,
	// or the whole file is rejected. Assert every required key on every element and that
	// result statuses stay within InSpec's ControlResultStatus enum.
	assert.Contains(t, legacy["platform"], "release", "platform.release is InSpec-required")
	profile := legacy["profiles"].([]any)[0].(map[string]any)
	for _, k := range []string{"attributes", "controls", "groups", "name", "sha256", "supports"} {
		assert.Contains(t, profile, k, "InSpec-required profile field %q must be present", k)
	}
	controls := profile["controls"].([]any)
	validStatus := map[string]bool{"passed": true, "failed": true, "error": true, "skipped": true}
	byID := map[string]map[string]any{}
	for _, cv := range controls {
		c := cv.(map[string]any)
		byID[c["id"].(string)] = c
		for _, k := range []string{"id", "impact", "refs", "results", "source_location", "tags"} {
			assert.Contains(t, c, k, "control %v missing InSpec-required %q", c["id"], k)
		}
		for _, rv := range c["results"].([]any) {
			r := rv.(map[string]any)
			assert.Contains(t, r, "code_desc")
			assert.Contains(t, r, "start_time")
			assert.Truef(t, validStatus[r["status"].(string)], "result status %q is not a valid InSpec ControlResultStatus", r["status"])
		}
	}

	// Waiver: control status flattened to the effective outcome (passed), the raw
	// result verdict preserved (failed), and an audit breadcrumb in waiver_data.
	waiver := byID["V-001-waiver"]
	assert.Equal(t, "passed", waiver["status"], "control status is the effective (attested) outcome")
	wd := waiver["waiver_data"].(map[string]any)
	assert.Equal(t, "waiver", wd["override_type"])
	assert.Equal(t, true, wd["skipped_due_to_waiver"])
	assert.Contains(t, wd["message"], "Compensating control")
	rawStatus := waiver["results"].([]any)[0].(map[string]any)["status"]
	assert.Equal(t, "failed", rawStatus, "raw per-result verdict is preserved")

	// False positive: also flattened, breadcrumb records the disposition type.
	fp := byID["V-002-fp"]
	assert.Equal(t, "passed", fp["status"])
	assert.Equal(t, "falsePositive", fp["waiver_data"].(map[string]any)["override_type"])

	// POA&M: not representable — control stays failed, breadcrumb records it, warning names it.
	poam := byID["V-003-poam"]
	assert.Equal(t, "failed", poam["status"])
	assert.Contains(t, poam["waiver_data"].(map[string]any), "not_representable_in_v2")

	// riskAdjustment: the effective (re-scored) impact is what the v2 control shows.
	risk := byID["V-004-risk"]
	assert.InDelta(t, 0.3, risk["impact"].(float64), 1e-9)
	// Incidental gap C: InSpec resource fields carried onto the result.
	riskResult := risk["results"].([]any)[0].(map[string]any)
	assert.Equal(t, "file", riskResult["resource_class"])
	assert.Equal(t, "/etc/audit/auditd.conf", riskResult["resource_id"])
	// refs carried into the v2 refs slot; cwe/severity mirrored into tags for Heimdall.
	assert.NotEmpty(t, risk["refs"], "advisory refs are carried")
	riskTags := risk["tags"].(map[string]any)
	assert.Equal(t, []any{"CWE-79"}, riskTags["cweid"])
	assert.Equal(t, "medium", riskTags["severity"])

	// Part B: the non-representable POA&M is surfaced as a warning, not dropped silently.
	joined := strings.Join(warnings, "\n")
	assert.Contains(t, joined, "V-003-poam")
	assert.Contains(t, joined, "POA&M")
}

// TestDowngradeV3ToV2_ValidatesAgainstInspecSchema is the authoritative regression
// guard: the downgrade output must validate against the InSpec exec-json schema
// Heimdall's parser enforces (vendored in testdata). Field-presence assertions can
// drift from the real contract; validating the whole document against the schema
// cannot.
func TestDowngradeV3ToV2_ValidatesAgainstInspecSchema(t *testing.T) {
	_, thisFile, _, _ := runtime.Caller(0)
	testdata := filepath.Join(filepath.Dir(thisFile), "testdata")
	input, err := os.ReadFile(filepath.Join(testdata, "modern_with_amendments.json"))
	require.NoError(t, err)

	out, _, err := TransformHDF(input, ModernVersion, LegacyVersion)
	require.NoError(t, err)

	// Load the schema from bytes rather than a file:// reference loader — a Windows
	// absolute path (D:\...) does not form a valid file URI and fails to parse.
	schemaBytes, err := os.ReadFile(filepath.Join(testdata, "exec-json.schema.json"))
	require.NoError(t, err)

	result, err := gojsonschema.Validate(
		gojsonschema.NewBytesLoader(schemaBytes),
		gojsonschema.NewBytesLoader(out),
	)
	require.NoError(t, err)
	for _, e := range result.Errors() {
		t.Errorf("downgraded document violates InSpec exec-json schema: %s", e)
	}
}

func TestDowngradeV3ToV2_SkipsExpiredOverrideBreadcrumb(t *testing.T) {
	// A requirement whose only override has already expired must not name that
	// override in the waiver_data breadcrumb.
	input := []byte(`{
		"baselines": [{"name": "B", "requirements": [{
			"id": "V-EXP", "title": "Expired waiver", "impact": 0.5,
			"effectiveStatus": "failed",
			"statusOverrides": [{
				"type": "waiver", "status": "passed", "reason": "old waiver",
				"appliedBy": {"type": "username", "identifier": "jdoe"},
				"appliedAt": "2019-01-01T00:00:00Z", "expiresAt": "2020-01-01T00:00:00Z"
			}],
			"results": [{"status": "failed", "codeDesc": "x", "startTime": "2020-01-01T00:00:00Z"}]
		}]}],
		"components": [{"name": "h", "type": "host"}],
		"generator": {"name": "g", "version": "1.0.0"},
		"tool": {"name": "t", "version": "1.0.0"}
	}`)
	out, _, err := TransformHDF(input, ModernVersion, LegacyVersion)
	require.NoError(t, err)

	var legacy map[string]any
	require.NoError(t, json.Unmarshal(out, &legacy))
	c := legacy["profiles"].([]any)[0].(map[string]any)["controls"].([]any)[0].(map[string]any)
	wd, _ := c["waiver_data"].(map[string]any)
	assert.NotContains(t, wd, "override_type", "an expired override must not become the waiver_data breadcrumb")
}

func TestDetectHDFVersion(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		want    string
		wantErr bool
	}{
		{"legacy (v2) with profiles+platform", `{"version":"3.4.5","profiles":[],"platform":{"name":"test"}}`, LegacyVersion, false},
		{"modern (v3) with baselines+components", `{"baselines":[],"components":[]}`, ModernVersion, false},
		{"ambiguous", `{"version":"1.0"}`, "", true},
		{"invalid json", `not json`, "", true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := DetectHDFVersion([]byte(tt.input))
			if tt.wantErr {
				assert.Error(t, err)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.want, got)
			}
		})
	}
}

func TestNormalizeVersion(t *testing.T) {
	// "1" has no distinct schema → maps to legacy (v2) with a warning.
	got, warn := NormalizeVersion("1")
	assert.Equal(t, LegacyVersion, got)
	assert.NotEmpty(t, warn, "hdf@1 should warn")

	// "2", "3", and "" pass through with no warning.
	for _, v := range []string{LegacyVersion, ModernVersion, ""} {
		got, warn := NormalizeVersion(v)
		assert.Equal(t, v, got)
		assert.Empty(t, warn, "hdf@%s should not warn", v)
	}

	// A leading "v" is accepted and stripped (users write "v3"/"v2").
	for _, v := range []string{"v2", "v3"} {
		got, warn := NormalizeVersion(v)
		assert.Equal(t, strings.TrimPrefix(v, "v"), got)
		assert.Empty(t, warn, "%s should normalize without warning", v)
	}
	// "v1" strips to "1" → legacy (v2) with the no-v1 warning.
	got, warn = NormalizeVersion("v1")
	assert.Equal(t, LegacyVersion, got)
	assert.NotEmpty(t, warn, "v1 should warn like 1")
}

File: hdf-converters/shared/typescript/hdf-version.test.ts
import { describe, it, expect } from 'vitest';
import { transformHDF, detectHDFVersion } from './hdf-version.js';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const fixturesDir = join(__dirname, '..', '..', 'converters', 'legacyhdf-to-hdf', 'fixtures', 'input');

function readFixture(name: string): string {
  return readFileSync(join(fixturesDir, name), 'utf-8');
}

describe('transformHDF', () => {
  it('upgrades v1 to v2', () => {
    const v1 = readFixture('minimal.json');
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed).toHaveProperty('baselines');
    expect(parsed).toHaveProperty('components');
    expect(parsed).not.toHaveProperty('profiles');
    expect(parsed).not.toHaveProperty('platform');
  });

  it('downgrades v2 to v1', () => {
    const v1 = readFixture('minimal.json');
    const v2 = transformHDF(v1, '1', '2');
    const v1Again = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1Again);
    expect(parsed).toHaveProperty('profiles');
    expect(parsed).toHaveProperty('platform');
    expect(parsed).not.toHaveProperty('baselines');
    expect(parsed).not.toHaveProperty('components');
  });

  it('returns input unchanged for same version', () => {
    const v1 = readFixture('minimal.json');
    const output = transformHDF(v1, '1', '1');
    expect(output).toBe(v1);
  });

  it('throws for unknown version pair', () => {
    expect(() => transformHDF('{}', '3', '2')).toThrow('No HDF transform');
  });

  it('round-trip preserves profile count', () => {
    const v1 = readFixture('minimal.json');
    const v2 = transformHDF(v1, '1', '2');
    const v1Again = transformHDF(v2, '2', '1');
    const original = JSON.parse(v1);
    const roundTripped = JSON.parse(v1Again);
    expect(roundTripped.profiles.length).toBe(original.profiles.length);
  });
});

describe('detectHDFVersion', () => {
  it('detects v1', () => {
    expect(detectHDFVersion('{"profiles":[],"platform":{"name":"test"}}')).toBe('1');
  });

  it('detects v2 with components', () => {
    expect(detectHDFVersion('{"baselines":[],"components":[]}')).toBe('2');
  });

  it('throws for ambiguous input', () => {
    expect(() => detectHDFVersion('{"version":"1.0"}')).toThrow('Cannot determine HDF version');
  });

  it('throws for invalid JSON', () => {
    expect(() => detectHDFVersion('not json')).toThrow();
  });
});

describe('transformHDF upgrade edge cases', () => {
  it('handles v1 with no platform', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'test',
        controls: [{
          id: 'ctrl-1',
          impact: 0.5,
          results: [{ status: 'passed', code_desc: 'ok', start_time: '2024-01-01T00:00:00Z' }],
        }],
      }],
      statistics: { duration: 1 },
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.components).toEqual([]);
    expect(parsed.baselines).toHaveLength(1);
  });

  it('handles v1 platform without release', () => {
    const v1 = JSON.stringify({
      platform: { name: 'test-host' },
      profiles: [],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.components[0].name).toBe('test-host');
    expect(parsed.components[0]).not.toHaveProperty('osVersion');
  });

  it('handles v1 with no profiles', () => {
    const v1 = JSON.stringify({ platform: { name: 'x', release: '1.0' } });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.baselines).toEqual([]);
    expect(parsed.statistics).toEqual({});
  });

  it('handles v1 control with no descriptions and no desc', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{ id: 'c1', impact: 0.7 }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const req = parsed.baselines[0].requirements[0];
    expect(req.descriptions).toEqual([{ label: 'default', data: '' }]);
  });

  it('handles v1 control with desc but no descriptions array', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{ id: 'c1', impact: 0.7, desc: 'My description' }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const req = parsed.baselines[0].requirements[0];
    expect(req.descriptions).toEqual([{ label: 'default', data: 'My description' }]);
  });

  it('handles v1 control with no title', () => {
    const v1 = JSON.stringify({
      profiles: [{ name: 'p', controls: [{ id: 'c1', impact: 0.5 }] }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.baselines[0].requirements[0]).not.toHaveProperty('title');
  });

  it('handles v1 control with no code or source_location', () => {
    const v1 = JSON.stringify({
      profiles: [{ name: 'p', controls: [{ id: 'c1', impact: 0.5 }] }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const req = parsed.baselines[0].requirements[0];
    expect(req).not.toHaveProperty('code');
    expect(req).not.toHaveProperty('sourceLocation');
  });

  it('handles v1 control with source_location having only ref', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{ id: 'c1', impact: 0.5, source_location: { ref: 'test.rb' } }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.baselines[0].requirements[0].sourceLocation).toEqual({ ref: 'test.rb' });
  });

  it('handles v1 control with source_location having only line', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{ id: 'c1', impact: 0.5, source_location: { line: 42 } }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.baselines[0].requirements[0].sourceLocation).toEqual({ line: 42 });
  });

  it('maps all v1 status values correctly', () => {
    const statuses = ['passed', 'failed', 'error', 'not_applicable', 'not_reviewed', 'skipped', 'custom_status'];
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: statuses.map((s, i) => ({
          id: `c${i}`,
          impact: 0.5,
          results: [{ status: s, code_desc: 'test', start_time: '2024-01-01T00:00:00Z' }],
        })),
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const resultStatuses = parsed.baselines[0].requirements.map(
      (r: Record<string, unknown>) => (r.results as Array<Record<string, unknown>>)[0].status,
    );
    expect(resultStatuses).toEqual([
      'passed', 'failed', 'error', 'notApplicable', 'notReviewed', 'notReviewed', 'custom_status',
    ]);
  });

  it('handles v1 result with message, exception, and backtrace', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{
          id: 'c1',
          impact: 0.5,
          results: [{
            status: 'failed',
            code_desc: 'should work',
            start_time: '2024-01-01T00:00:00Z',
            message: 'expected true got false',
            exception: 'AssertionError',
            backtrace: ['line1', 'line2'],
          }],
        }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const result = parsed.baselines[0].requirements[0].results[0];
    expect(result.message).toBe('expected true got false');
    expect(result.exception).toBe('AssertionError');
    expect(result.backtrace).toEqual(['line1', 'line2']);
  });

  it('handles v1 result with no code_desc or start_time', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{
          id: 'c1',
          impact: 0.5,
          results: [{ status: 'passed' }],
        }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const result = parsed.baselines[0].requirements[0].results[0];
    expect(result.codeDesc).toBe('');
    expect(result.startTime).toBeTruthy(); // defaults to current time
  });

  it('handles v1 result with empty backtrace', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        controls: [{
          id: 'c1',
          impact: 0.5,
          results: [{ status: 'passed', backtrace: [] }],
        }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const result = parsed.baselines[0].requirements[0].results[0];
    expect(result).not.toHaveProperty('backtrace');
  });

  it('handles profile with all optional metadata fields', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        title: 'My Title',
        version: '2.0',
        maintainer: 'MITRE',
        summary: 'A summary',
        license: 'Apache-2.0',
        copyright: 'MITRE 2024',
        copyright_email: 'test@mitre.org',
        groups: [{ id: 'grp1', title: 'Group 1', controls: ['c1'] }],
        controls: [],
        depends: [{ name: 'dep1' }],
        supports: [{ platform: 'linux' }],
        attributes: [{ name: 'attr1', default: 'val1' }],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    const bl = parsed.baselines[0];
    expect(bl.title).toBe('My Title');
    expect(bl.version).toBe('2.0');
    expect(bl.maintainer).toBe('MITRE');
    expect(bl.summary).toBe('A summary');
    expect(bl.license).toBe('Apache-2.0');
    expect(bl.copyright).toBe('MITRE 2024');
    expect(bl.copyrightEmail).toBe('test@mitre.org');
    expect(bl.groups).toEqual([{ id: 'grp1', title: 'Group 1', requirements: ['c1'] }]);
    expect(bl.depends).toEqual([{ name: 'dep1' }]);
    expect(bl.supports).toEqual([{ platform: 'linux' }]);
    expect(bl.inputs).toEqual([{ name: 'attr1', default: 'val1' }]);
  });

  it('handles group with no title', () => {
    const v1 = JSON.stringify({
      profiles: [{
        name: 'p',
        groups: [{ id: 'grp1' }],
        controls: [],
      }],
    });
    const v2 = transformHDF(v1, '1', '2');
    const parsed = JSON.parse(v2);
    expect(parsed.baselines[0].groups[0]).toEqual({ id: 'grp1', requirements: [] });
    expect(parsed.baselines[0].groups[0]).not.toHaveProperty('title');
  });
});

describe('transformHDF downgrade edge cases', () => {
  it('handles v2 with no components', () => {
    const v2 = JSON.stringify({
      baselines: [],
      components: [],
      statistics: {},
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    expect(parsed.platform.name).toBe('');
  });

  it('handles v2 component with osVersion and name', () => {
    const v2 = JSON.stringify({
      baselines: [],
      components: [{ name: 'my-host', type: 'host', osVersion: '22.04' }],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    expect(parsed.platform.name).toBe('my-host');
    expect(parsed.platform.release).toBe('22.04');
    expect(parsed.platform.target_id).toBe('my-host');
  });

  it('handles v2 component with no osVersion', () => {
    const v2 = JSON.stringify({
      baselines: [],
      components: [{ name: 'my-host' }],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    expect(parsed.platform.name).toBe('my-host');
    expect(parsed.platform).not.toHaveProperty('release');
  });

  it('handles v2 component with no name', () => {
    const v2 = JSON.stringify({
      baselines: [],
      components: [{ type: 'host' }],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    expect(parsed.platform.name).toBe('');
    expect(parsed.platform).not.toHaveProperty('target_id');
  });

  it('handles v2 baseline with optional fields', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        version: '1.0',
        title: 'Baseline',
        maintainer: 'MITRE',
        summary: 'Sum',
        license: 'MIT',
        copyright: 'MITRE',
        copyrightEmail: 'test@test.com',
        groups: [{ id: 'g1', title: 'Group', requirements: ['r1'] }],
        requirements: [],
        depends: [{ name: 'dep' }],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const prof = parsed.profiles[0];
    expect(prof.name).toBe('bl');
    expect(prof.version).toBe('1.0');
    expect(prof.title).toBe('Baseline');
    expect(prof.maintainer).toBe('MITRE');
    expect(prof.summary).toBe('Sum');
    expect(prof.license).toBe('MIT');
    expect(prof.copyright).toBe('MITRE');
    expect(prof.copyright_email).toBe('test@test.com');
    expect(prof.groups[0].controls).toEqual(['r1']);
  });

  it('handles v2 baseline with no optional fields', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const prof = parsed.profiles[0];
    expect(prof.name).toBe('bl');
    expect(prof).not.toHaveProperty('version');
    expect(prof).not.toHaveProperty('title');
  });

  it('handles v2 requirement with all status types in downgrade', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [
          {
            id: 'r1',
            impact: 0.5,
            descriptions: [{ label: 'default', data: 'desc' }],
            results: [
              { status: 'passed', codeDesc: 'ok', startTime: '2024-01-01T00:00:00Z' },
              { status: 'failed', codeDesc: 'bad', startTime: '2024-01-01T00:00:00Z' },
              { status: 'error', codeDesc: 'err', startTime: '2024-01-01T00:00:00Z' },
              { status: 'notApplicable', codeDesc: 'na', startTime: '2024-01-01T00:00:00Z' },
              { status: 'notReviewed', codeDesc: 'nr', startTime: '2024-01-01T00:00:00Z' },
              { status: 'custom', codeDesc: 'cust', startTime: '2024-01-01T00:00:00Z' },
            ],
          },
        ],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const ctrl = parsed.profiles[0].controls[0];
    const statuses = ctrl.results.map((r: Record<string, unknown>) => r.status);
    expect(statuses).toEqual(['passed', 'failed', 'error', 'not_applicable', 'not_reviewed', 'custom']);
  });

  it('handles v2 requirement with sourceLocation in downgrade', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [{
          id: 'r1',
          title: 'Req 1',
          impact: 0.5,
          code: 'control "r1" do end',
          sourceLocation: { ref: 'test.rb', line: 10 },
          descriptions: [{ label: 'default', data: 'desc' }],
          results: [],
        }],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const ctrl = parsed.profiles[0].controls[0];
    expect(ctrl.source_location).toEqual({ ref: 'test.rb', line: 10 });
    expect(ctrl.code).toBe('control "r1" do end');
    expect(ctrl.desc).toBe('desc');
  });

  it('handles v2 requirement with no desc in descriptions', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [{
          id: 'r1',
          impact: 0.5,
          descriptions: [{ label: 'rationale', data: 'why' }],
          results: [],
        }],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const ctrl = parsed.profiles[0].controls[0];
    expect(ctrl).not.toHaveProperty('desc');
  });

  it('handles v2 result with message, exception, backtrace', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [{
          id: 'r1',
          impact: 0.5,
          descriptions: [],
          results: [{
            status: 'failed',
            codeDesc: 'test',
            startTime: '2024-01-01T00:00:00Z',
            runTime: 0.5,
            message: 'msg',
            exception: 'exc',
            backtrace: ['a', 'b'],
          }],
        }],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const result = parsed.profiles[0].controls[0].results[0];
    expect(result.run_time).toBe(0.5);
    expect(result.message).toBe('msg');
    expect(result.exception).toBe('exc');
    expect(result.backtrace).toEqual(['a', 'b']);
  });

  it('handles v2 result with no startTime', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [{
          id: 'r1',
          impact: 0.5,
          descriptions: [],
          results: [{ status: 'passed', codeDesc: 'ok' }],
        }],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    const result = parsed.profiles[0].controls[0].results[0];
    expect(result).not.toHaveProperty('start_time');
  });

  it('handles v2 group with no title in downgrade', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        groups: [{ id: 'g1', requirements: ['r1'] }],
        requirements: [],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    expect(parsed.profiles[0].groups[0]).toEqual({ id: 'g1', controls: ['r1'] });
    expect(parsed.profiles[0].groups[0]).not.toHaveProperty('title');
  });

  it('handles sourceLocation with ref but no line in downgrade', () => {
    const v2 = JSON.stringify({
      baselines: [{
        name: 'bl',
        requirements: [{
          id: 'r1',
          impact: 0.5,
          descriptions: [],
          sourceLocation: { ref: 'test.rb' },
          results: [],
        }],
      }],
      components: [],
    });
    const v1 = transformHDF(v2, '2', '1');
    const parsed = JSON.parse(v1);
    expect(parsed.profiles[0].controls[0].source_location).toEqual({ ref: 'test.rb' });
  });
});
